app/models/manageiq/providers/inventory/persister/builder/shared.rb

```diff
@@ -72,9 +72,31 @@ def hardwares
     end
 
     def operating_systems
+      custom_save_block = lambda do |_ems, inventory_collection|
+        vms_and_templates_ids = inventory_collection.data.map { |os| os.vm_or_template&.id }.compact
+
+        vms_and_templates_index = VmOrTemplate.where(:id => vms_and_templates_ids).index_by(&:id)
+        operating_systems_index = OperatingSystem.where(:vm_or_template_id => vms_and_templates_ids).index_by(&:vm_or_template_id)
+
+        inventory_collection.each do |inventory_object|
+          vm_or_template = vms_and_templates_index[inventory_object.vm_or_template.id]
+
+          # If a VM has had smartstate run on it (drift_states are present) then don't overwrite the
+          # operating system record with one from the provider.  This is because typically far more
+          # details and correct information can be found from smartstate.
+          next unless vm_or_template.drift_states.blank? || vm_or_template.operating_system.nil? ||
+                      vm_or_template.operating_system.product_name.blank?
```

Comment:
Logic lifted from https://github.com/ManageIQ/manageiq/blob/master/app/models/ems_refresh/save_inventory.rb#L216

Reply:
These are all N+1 queries without some preloads

Reply:
Yeah, how about
`DriftState.where(:resource_type => "VmOrTemplate", :resource_id => vms_and_templates_ids).select(:resource_id).distinct.index_by(&:resource_id)`

with hundreds of drift states on multiple vms that gives me:
```
>> vms_and_templates_ids = [1, 2, 3, 4]
=> [1, 2, 3, 4]
>> DriftState.where(:resource_type => "VmOrTemplate", :resource_id => vms_and_templates_ids).select(:resource_id).distinct.index_by(&:resource_id)
  DriftState Load (0.8ms)  SELECT DISTINCT "drift_states"."resource_id" FROM "drift_states" WHERE "drift_states"."resource_type" = $1 AND "drift_states"."resource_id" IN (1, 2, 3, 4)  [["resource_type", "VmOrTemplate"]]
  DriftState Inst Including Associations (0.1ms - 3rows)
=> {1=>#<DriftState id: nil, resource_id: 1>, 2=>#<DriftState id: nil, resource_id: 2>, 4=>#<DriftState id: nil, resource_id: 4>}
```

Reply:
Done